### ISSUE TITLE
Boolean and real-world usage [casting annotation]

### ISSUE THREAD

**Issue by Mouvedia** [ericelliott/rtype#73]
Quite often when you are using the Boolean [builtin type](https://github.com/ericelliott/rtype#builtin-types) you just mean that you _expect_ a primitive value (`true` or `false`) even if anything could be passed.
How could we annotate that **preference**? Today we would use `Any`; but I would like to be able to clearly state that an implicit (or explicit) Boolean casting will be performed regardless of what you pass.

Here's an example of what we could have:

```
toggle(String, ?: Boolean(Any)) => Boolean

// would work the same for the other castable builtin types
signatureName(Number(CustomType)) => Number
```

**Comment by ericelliott** [comment]
> Quite often when you are using the Boolean builtin type you just mean that you expect a primitive value (true or false) even if anything could be passed.

Use `Any` if you'll take any value. If you really mean boolean, let the user explicitly cast before calling.

The signature should not reveal the internal implementation of your function. You state the input types and the output types... you don't say what you do with those values in-between.

**Comment by Mouvedia** [comment]
@ericelliott I have a function that accepts anything as an argument and it will be coerced to a boolean. The problem with `Any` is that the intent—an essential role of the documentation—is unclear. In short the user reading the documentation doesn't gain any insight into what the argument might be used for.

Here's what I came up with:

```
String.parseAsClass(class: String, HTML5?: !!Any)
    => Void,
    throws TypeError | SyntaxError | DOMException,
    requires: String.isString, String.prototype.inspect
```

My argument is that no one really checks using both `=== true` **and** `=== false` for each path. If I managed to convince you feel free to reopen to discuss the syntax.

**Comment by ericelliott** [comment]
Not only am I not convinced we should tell the users that we're casting to a different type internally, I think it's a much better practice to simply ask for a Boolean to begin with, partially because I agree with this -- if you're ambiguous about the type you're really looking for:

> the intent—an essential role of the documentation—is unclear.

**Comment by Mouvedia** [comment]
Then you are asking for Boolean but still accepting anything internally as if it was considered an undocumented feature. I think this is worth mentioning in some way and I will in my documentation even if I end up with an incompatible rtype variant.

**Comment by ericelliott** [comment]
I'm curious... why don't you simply ask for a boolean if a boolean is what you want?

**Comment by Mouvedia** [comment]
Well I am not. Well anyway I don't think I am the only one thinking it's warranted* but maybe it doesn't belong in the MVP. The solution I was proposing would have been cryptic and it wouldn't be on par with the rest of the specification.

\* https://github.com/ericelliott/rtype/issues/89#issuecomment-210981207

**Comment by ericelliott** [comment]
The following is all true of Boolean... which leaves the open question, if `Boolean` does what you want, why use the less-specific `BooleanMixed` or whatever we'd name it? 

``` js
var foo: %Boolean%%Mixed = 2;
!foo; // legal
if (foo) {} // legal

foo.bar; // illegal, type Boolean does not have field bar.
```

You'd have to explain in your docs what `BooleanMixed` means, because it's not a type that users will be automatically familiar with. By the time you're done explaining the awkwardness, the user could have simply passed a `Boolean` and moved on a lot faster...

So, would it _really_ be convenient to introduce such a type when you really only want a `Boolean` internally?

**Comment by Mouvedia** [comment]
> …it's not a type that users will be automatically familiar with.

Exactly the point I was raising. I see no elegant solution to my problem.

> …when you really only want a Boolean internally?

Again, no I don't want Boolean; I want Any—which gets coerced to to a boolean.

**Comment by ericelliott** [comment]
> I don't want Boolean; I want Any—which gets coerced to to a boolean.

This is the part I don't understand. If it gets coerced to a boolean, why do you want `Any`? What is the advantage of allowing any type to be passed in when you're treating it only as a `Boolean`?

**Comment by Mouvedia** [comment]
There's a pattern that comes back very often where you use the existence of something to switch a boolean. This is something that is ubiquitous with most web APIs.

e.g. `Element.classList.toggle` accepts `Any` as his force argument and rightly so: it is expected.

**Comment by ericelliott** [comment]
Perhaps this was just a bad example, but IMO, you've just highlighted why it's important to be clear about the type you're actually expecting. `Element.classList.toggle`'s `force` parameter expects a `Boolean`, and if you pass it something else, it may happen to work in one or two browsers, but it definitely won't work in all environments.

For example, if you feed the [canonical shim](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) a truthy value that is not `=== true`, you'll get unexpected results.

Granted:

> where you use the existence of something to switch a boolean

This is a common pattern, but if that something is being passed unchanged to other processes, what you really need is `Any`, because your function is not the deciding factor in what that type should be.

If that something is not exported from your function unchanged, what you really need is to be clear that you're expecting a `Boolean` value in order to avoid mistakes like assuming that you can pass whatever into functions like `Element.classList.toggle`.

From the HTML spec:

```
The toggle(token, force) method, when invoked, must run these steps:

1. If token is the empty string, then throw a SyntaxError.

2. If token contains any ASCII whitespace, then throw an InvalidCharacterError.

3. If token is in tokens, run these substeps:

    1. If force is either not passed or is false, then remove token from tokens, run the update steps, and return false.

    2. Otherwise, return true.

4. Otherwise, run these substeps:

    1. If force is passed and is false, return false.

    2. Otherwise, append token to tokens, run the update steps, and return true.
```

Note that according to the spec, `force` is checked for `false`, not for falsey values.

**Comment by Mouvedia** [comment]
I knew about the shim's problem. That's a bug IMHO because switching a class based on the existence of a node—or anything dynamic by nature—is a feature: you shouldn't have to use a ternary there. Since [the `force` argument wasn't really well supported](https://connect.microsoft.com/IE/feedback/details/878564/element-classlist-toggle-does-not-support-second-parameter), it doesn't matter.

Anyhow, the issue is more about separation of concerns and the scope of rtype than whether there is case to be made. In a dynamic language like JS, I think it would be useful. I didn't manage to persuade you: Ill leave it there.

**Comment by ericelliott** [comment]
> That's a bug IMHO

Not according to the spec. You really shouldn't be passing `Any` values into that function... you're essentially taking advantage of undocumented features of particular browser implementations, which means that your code may break on browsers you're not testing on. Does it work in Microsoft Edge, for example? What if a new browser gets introduced to the market and steals 80% user share and complies with the spec's `=== false` rules? Then all the software you passed `Any` values in will be broken. By the way, this has happened to me before using different HTML APIs, which forced me to work late on a weekend to fix broken shopping carts.

Just because something happens to work doesn't mean it's a good idea.

> Anyhow, the issue is more about separation of concerns and the scope of rtype

Yeah, while I'm aware it's common practice to take `Any` types and coerce them to booleans inside your functions, what you really need for those cases is `Any` or to be more restrictive and call for `Boolean`. Which one is the right call depends on your function:
1. Does the function export the value or use the additional capabilities of the type passed in? If so, you need `Any`, and `Boolean` won't do the job you need it to do.
2. Does the function simply coerce to `Boolean`? If so, you should be clear that you're expecting a boolean value. Doing so will make your API more clear for users calling your function. This may force them to coerce values to Boolean themselves, but `!!foo` is hardly a challenge to type. You're literally saving the caller 2 keystrokes. Is it worth the ambiguity and potential confusion?

**Comment by Mouvedia** [comment]
I don't want to argument about something off topic. Here's a link to the [specification](https://dom.spec.whatwg.org/#dom-domtokenlist-toggle). The relevant word there is "Otherwise". It was just an example, you probably have done that kind of things before.

> Just because something happens to work doesn't mean it's a good idea.

There's a reason behind that ubiquity on the browser side of things. _Every time_ that you have a pseudo boolean argument you can pass anything. It's useful in that context. Now I am against having boolean arguments for the reasons outlined in that [conference](https://www.youtube.com/watch?v=y7RrP_fOYUQ&feature=youtu.be&t=640). But if we are past that then I would never expect it to throw if I didn't expressly pass `true` or `false`.

> Is it worth the ambiguity and potential confusion?

I don't know.

**Comment by ericelliott** [comment]
> The relevant word there is "otherwise".

That's even worse. It seems at first glance that it should coerce values to boolean, but _that's not what happens._ What really happens (according to the spec that I quoted verbatim above) is an `=== false` check, which means:

If you pass `0`, `null`, `NaN`, an empty string, it behaves as if you passed `true`. Incidentally, this is probably why the polyfill checks for `=== true`, instead.

> Every time that you have a ~boolean~ argument you can pass anything.

As we just demonstrated with the `Element.classList.toggle` function, that simply isn't true.

> I would never expect it to throw if I didn't expressly pass true or false.

It seems to me that you're writing time bombs in your code because a function that should throw when you don't pass a `Boolean` value doesn't throw, and that gave you a false impression of how the function behaves.

**Comment by Mouvedia** [comment]
Frankly I don't know if they meant the primitive booleans `true` and `false` or the general concept of truthy and falsy. Everything is so generic in the specification it's often hard to tell. No matter how many examples Id give you that wouldn't help in the least :)

> …because a function that should throw…

We disagree on that part. Not everyone has the same expectations. It's safer to err on the side of restriction, isolation and least effort; that explains your call.
